### PR TITLE
Added unit tests and made few minor changes to YANN code

### PR DESCRIPTION
**Pull request by buddhapuneeth**
1. Added unit tests to utils, core, modules 
2. Restructured tests directory to make it inline with Yann directory
3. Made below changes to YANN:
    a. added 'pragma:no cover' in places like __main__ which can't be covered.
    b. added 'pragma: no cover' in mixedout condition in maxout() of activation.py which is yet to implement completely.
        Also added in the internal 'else' conditions of type == 'maxout' and type == 'meanout' as maxout_out variable will be always 'none'
    c. added 'pragma: no cover' in conv_pool.py where codecov is considering commented code which ideally should be ignored
    d. added code in merge.py to accommodate dropout and changed conditional check.
    e. added code in raster.py: changed from x.dtype to x[i].dtype, as x is a tuple.
    f. added code in image.py: imported scipy, changed if channel = 1 the parameters of transpose from 4 to 3.
    e. added code in dataset.py: In load_mat_data() removed the default value for location which was pointing to waldo

### CODE REVIEW

**Comment by ragavvenkatesan** [comment]
Can you find out why the commented section isn't working ? in convpool.py.

File: yann/utils/image.py [hunk 1]
```diff
@@ -45,7 +51,7 @@ def gray2rgb(r, g, b, channels_dim = 3):
     if channels_dim == 3:
         return out
     elif channels_dim == 1:
-        return out.transpose((0,3,1,2))
+        return out.transpose((2,0,1))
```
Comment (ragavvenkatesan) [inline yann/utils/image.py]: Are you sure of this change ? This feels wrong to me. Can you verify this by creating a cifar 10 dataset and running it with pre_process rgb and then see how the visualizer prints down the data as images ?

File: yann/utils/raster.py [hunk 2]
```diff
@@ -98,7 +98,7 @@ def tile_raster_images(X, img_shape, tile_shape, tile_spacing=(0, 0),
                                     dtype='uint8')
         else:
             out_array = numpy.zeros((out_shape[0], out_shape[1], 4),
-                                    dtype=X.dtype)
+                                    dtype=X[1].dtype)
```
Comment (ragavvenkatesan) [inline yann/utils/raster.py]: Again, are you sure of this ? Check this too please on an actual visualizer code.

**Comment by buddhapuneeth** [comment]
In convpool, below section was already commented. We just added 'no cover' statement as codecov tool is considering that code also. We are working with codecov and Travis developers on few such issues.

 I think there is no pooling concept for deconv layers, that's why you are throwing an exception.

     """ #pragma: no cover
 pooler = pooler_2d(
 input = conv_out,
 img_shp = conv_out_shp,
 mode = pooltype,
 ds = poolsize,
 verbose = verbose
 )
 pool_out = pooler.out
 pool_out_shp = pooler.out_shp
 """

File: yann/utils/raster.py (hunk 2 above)
Comment (amotupal) [inline yann/utils/raster.py]: Yeah, If you check a couple of lines before that you are checking if X is a tuple "if isinstance(X, tuple)". This code will be executed only if X is a tuple, when it does it's throwing an error that tuple don't have a dtype. That's why I used that X[1] for dtype.

File: yann/utils/image.py (hunk 1 above)
Comment (amotupal) [inline yann/utils/image.py]: The code couple of lines before this asserts if the number of dimensions in r,g,b are 2. We are adding another dimension to it by concatenating them. In total we have 3 dimension with channels as the third one with index 2 ie., B01C format. So to get the channel at 1 i.e., convert it to BC01 format. we need to do give 2,0,1 to transpose such that the first dimension is channels.

**Pull request merged**